### PR TITLE
KAFKA-12558: Do not prematurely mutate partiton state and provide con…

### PR DESCRIPTION
**Pull request by emilnkrastev**
This PR addresses the issue described here [KAFKA-12558](https://issues.apache.org/jira/browse/KAFKA-12558). 

Additionally, The PR also allows to configure the max outstanding syncs in MirrorSourceTask because it is currently [hardcoded](https://github.com/apache/kafka/blob/trunk/connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java#L53).
A lot of offset syncs messages are lost during burst of messages in the source cluster or when the MirrorMaker has a lot to catch up (fist run or being inactive for a while). In such scenario it will take a while to sync the offsets in the destination cluster with partitions without regular activity even with reaching the maximum parallelism - 1 task per partition.

The PR tries to mitigate the issue by providing a way to change the maximum allowed concurrent offset syncs so that there are less offset syncs loses.

Here are my steps to reproduce the offset syncs issue because of the max outstanding syncs limited to 10:
1. Source topic with 12 partitions and 1400 messages with minimal activity. Messages are getting produced on daily basis
2. Run MirrorMaker2 process within the destination cluster network with offset syncs topic location set to target and 5 tasks
3. 372 offset syncs messages arrived in the destination cluster offset syncs topic
4. 9 out of 12 partitions are not synced correctly in the destination cluster
5. Waiting for hours and more for new messages to arrive in source Kafka cluster which will sync the correct offsets

### CODE REVIEW

**Comment by gharris1727** [comment]
@emilnkrastev Thanks for the fix! Are you still interested in making this change?

If so, I would suggest removing the configuration changes, as those would require a KIP. We can much more easily merge a targeted bug fix.

Also I noticed that your name and email are not set in your commit. You may want to configure these so that the change can be attributed to you.

Thanks!

**Comment by emilnkrastev** [comment]
@gharris1727 I'm still interested in the change and I will update the MR in the next couple of days.

Thanks for the reply!

**Comment by emilnkrastev** [comment]
@gharris1727 The PR is updated. Could you please take a look?

**Comment by gharris1727** [comment]
Thanks for the update @emilnkrastev.

It appears there are some CI failures that mention offsets translation that we need to look into, especially in `MirrorConnectorsIntegrationSSLTest`

Additionally, I'm trying to think of ways to reliably test this behavior. You've translated the existing test to work with the new update signature, but we haven't really targeted the semaphore balking behavior in a regression test. The lack of code coverage is partly why we didn't notice this before.

Here's a couple of ideas that might work, i'll let you choose whichever one you think makes more sense.
* Mock the semaphore acquire/release and unit test to simulate balking in MirrorSourceTaskTest
* Mock the producer and delay the send callback to cause the real semaphore to balk in MirrorSourceTaskTest
* Refactor the MirrorSourceTask semaphore to make it more unit-testable.
* Add a test to MirrorConnectorsIntegrationTestBase with a large contention on this semaphore and assert that the checkpoint offsets are sensible.

Thanks!

**Comment by emilnkrastev** [comment]
@gharris1727 The PR has been updated by adding additional unit test for testing partition state mutation and changing the update logic. The MirrorConnectorsIntegrationSSLTest Integration test was failing due to missing synced offsets and it was fix by always updating the previous upstream/downstream offset (the original logic is doing it).

There are still test failures but I believe there are not related to the PR changes. I can see similar test failures on the other PR's CI pipelines.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java
Comment (gharris1727) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java:308]: Nice change!

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java
Comment (gharris1727) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java]: nit: extra line

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java
Comment (gharris1727) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java]: nit
```suggestion
        assertEquals(newRecordOffset, partitionState.previousUpstreamOffset, "update previous upstream offset");
        assertEquals(newMetadataOffset, partitionState.previousDownstreamOffset, "update previous upstream offset");
```

**Review by gharris1727 — APPROVED** [review]
Thanks so much @emilnkrastev, this LGTM once some small nits are resolved.
The test looks great, and I'm glad that we're actually covering this code path. If all of the other tests are fine with those three instance variables being null, I'm wondering how much test coverage we're really missing right now.

**Comment by gharris1727** [comment]
@C0urante could you take a look at this, please?

**Comment by C0urante** [comment]
Thanks for this fix, @emilnkrastev. I want to make sure I understand the goal here before commenting on the actual code changes (although I've made sure to read through them carefully before writing this).

It looks like incorrect updates to the `partitionStates` field (or its contents) might result in the starvation of offset syncs for some topic partitions. The task will believe that it has performed syncs that it hasn't, and as a result, may never see that the lag between the (true) last-synced offset and the offset for the last-replicated record has exceeded the value for the user-configured `offset.lag.max` property.

However, it looks like KAFKA-12468 and possibly KAFKA-12558 describe a different problem: it's not just a matter of offset syncs being delayed for longer than expected or even indefinitely; instead, the offset syncs that do make it to the downstream cluster actually contain incorrect values, which in some extreme cases can result in things like negative consumer lag.

I don't see how this PR would address an issue with correctness in the actual offsets used for syncs; instead, it seems like the value here is that it will allow offset syncs to take place in cases where they should be but are currently not.

Is that a fair assessment of the goal here? If so, I'm happy to review/merge a fix with that goal, and if not, would you mind shedding some light on what the goal is and where I might be missing something?

**Comment by emilnkrastev** [comment]
@C0urante your assessment of the goal is correct - the PR changes are aiming to allow syncs to take place where they should be.

I believe there are 2 different problems - one that results in delayed offset syncs and the other that results in negative lag in the destination cluster. I'm trying to mitigate the first problem (delayed offset syncs).

**Comment by C0urante** [comment]
Thanks @emilnkrastev, good to know we're on the same page!

Thinking about this a little more, I wonder if a slightly more-involved approach is warranted. With the current proposal, it looks like we might still miss some cases that are predicated on the last-seen (not last-synced) upstream/downstream offsets.

For example, if the downstream topic is deleted and then recreated, the `downstreamOffset < previousDownstreamOffset` part of the [condition in `shouldSyncOffsets`](https://github.com/emilnkrastev/kafka/blob/67f6f71eef29b474ae642449de73862bf65aef7e/connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java#L302) will evaluate to `true`, and we'll try to send an offset sync as a result. If there are too many in-flight offset syncs already and we can't acquire access to the semaphore, we'll skip that sync--but since we'll also have already updated the `previousDownstreamOffset` field, the next call to `shouldSyncOffsets` won't (necessarily) return `true`, even though we should still attempt a sync in that case.

For a short-term fix, I think the `PartitionState` class could be adapted to "remember" whether syncs are necessary, and allow external callers to clear that state whenever an offset sync has actually been performed. That way, we never run the risk of "dropping" offset syncs that were supposed to be performed but were blocked by access to the semaphore.

How does that sound?

**Comment by emilnkrastev** [comment]
@C0urante Your suggestion sounds perfect.

I have updated the PR.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java
Comment (C0urante) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:73]: I'm not sure we should be mocking the semaphore here, since it's a bit of an implementation detail and I've already proposed that we split it out into several different semaphores (one for each topic partition) in https://github.com/apache/kafka/pull/12366.

Can we get the same testing coverage/guarantees by mocking out the offset sync producer instead, and selectively invoking/not invoking the callbacks passed to it in `send`?

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java
Comment (C0urante) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:307]: I was wondering if we should also move the updates for the `lastSyncUpstreamOffset` and `lastSyncDownstreamOffset` fields to this method, but there doesn't appear to be any downsides to this approach given our current usages of `update` and `reset`. 👍

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java
Comment (C0urante) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java:96]: Is it worth adding coverage to this case (and possibly `testZeroOffsetSync`) for when we invoke `update` repeatedly without invoking `reset`?

**Review by C0urante — COMMENTED** [review]
Thanks @emilnkrastev! The functional parts LGTM; just a few thoughts about the testing strategy.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java
Comment (emilnkrastev) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java:96]: I increased the test coverage.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java
Comment (emilnkrastev) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:73]: I have updated the MR without mocking the outstanding syncs semaphore but I have preserved the semaphore constructor parameter. The semaphore is getting initialized in the `start` method and I could not find a better way to avoid the NPE.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java
Comment (C0urante) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java]: This can be simplified a bit by using an `ArgumentCaptor`:

```suggestion
        ArgumentCaptor<Callback> producerCallback = ArgumentCaptor.forClass(Callback.class);
        when(producer.send(any(), producerCallback.capture())).thenAnswer(mockInvocation -> {
            producerCallback.getValue().onCompletion(null, null);
            return null;
        });
```

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java
Comment (C0urante) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java]: Using the `ArgumentCaptor` approach:

```suggestion
        producerCallback.getValue().onCompletion(null, null);
```

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java
Comment (C0urante) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java]: Using the `ArgumentCaptor` approach:
```suggestion
        // Do not immediately release the outstanding sync semaphore
        doReturn(null).when(producer).send(any(), producerCallback.capture());
```

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java
Comment (C0urante) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java]: I don't think we need coverage this fine-grained here; we get plenty of guarantees for the internal state tracking of the `PartitionState` class in the `testOffsetSync` and `testZeroOffsetSync` cases. Adding more here is going to make the test more brittle, and IMO the tradeoff isn't worth it.

All I'd want to see out of a test case like this is this series of events:
1. Cause offset sync to take place by invoking `commitRecord`
2. Invoke producer callback used for step 1 to signal to the task that the offset sync succeeded
3. Cause another offset sync to take place by invoking `commitRecord`
4. Do not invoke the producer callback used for step 3 (yet)
5. Cause another offset sync to take place, and verify that no calls to `Producer::send` were made as a result
6. Invoke the producer callback used for step 3
7. Cause another offset sync to take place, and verify that `Producer::send` was invoked as a result

We can optionally verify the actual content of the offset syncs (i.e., upstream offset/downstream offset/topic partition).

Other details like the assertions on the values of any `PartitionState` fields can and should be omitted.

**Review by C0urante — COMMENTED** [review]
Thanks @emilnkrastev! One final round of comments and this should be good to go.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java
Comment (C0urante) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceTask.java:73]: One alternative is to hardcode the instantiation of the `outstandingOffsetSyncs` semaphore in the testing-only constructor, without relying on a constructor parameter (and the same can be done for the `partitionStates` field).

But that has its own problems of duplicating instantiation logic for those fields and leading to possible divergence in behavior between task instances depending on whether they're brought up for testing or not, if we're not careful about keeping that instantiation logic in sync.

One way that that issue can be addressed is by pulling out any non-trivial instantiation logic into a separate static method; e.g.:
```java
private static Semaphore newOutstandingOffsetSyncsSemaphore() {
    return new Semaphore(MAX_OUSTANDING_OFFSET_SYNCS);
}
```

But I don't think that this is so much better than what you have right now that I'd block the PR on adapting this approach; even if it has some advantages, the additional complexity makes it debatable whether it'd really be worth the tradeoff.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java
Comment (emilnkrastev) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java]: Updated.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java
Comment (emilnkrastev) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java]: Updated.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java
Comment (emilnkrastev) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java]: Updated.

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java
Comment (emilnkrastev) [reply connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorSourceTaskTest.java]: Updated.

**Review by C0urante — APPROVED** [review]
LGTM, thanks @emilnkrastev!

**Pull request merged**
